### PR TITLE
Add project/duplicate endpoint

### LINKED ISSUE #10367: `duplicate project` endpoint for local project manager
Similarly to a cloud version, a local project manager should allow for duplicating an existing project.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #10367

Changelog:
- add: `project/duplicate` project manager command to duplicate existing projects

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: docs/language-server/protocol-project-manager.md [hunk 1]
```diff
@@ -42,6 +42,7 @@ transport formats, please look [here](./protocol-architecture.md).
   - [`project/delete`](#projectdelete)
   - [`project/listSample`](#projectlistsample)
   - [`project/status`](#projectstatus)
+  - [`project/duplicate`](#projectduplicate)
```
Comment (JaroslavTulach) [inline docs/language-server/protocol-project-manager.md:45]: Should it be called duplicate? Why not `copy` or `fork`?

File: docs/language-server/protocol-project-manager.md [hunk 2]
```diff
@@ -750,6 +751,50 @@ interface ProjectStatusResponse {
...
+- **Visibility:** Public
+
+#### Parameters
+
+```typescript
+interface ProjectDuplicateRequest {
+  /**
+   * The project to duplicate.
+   */
+  projectId: UUID;
```
Comment (JaroslavTulach) [inline docs/language-server/protocol-project-manager.md:770]: Shouldn't the message support (optional) suggested name for the newly created project?

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/file/BlockingFileSystem.scala [hunk 3]
```diff
@@ -89,6 +89,20 @@ class BlockingFileSystem[F[+_, +_]: Sync: ErrorChannel](
       }
       .mapError(toFsFailure)
 
+  /** @inheritdoc */
+  override def copy(from: File, to: File): F[FileSystemFailure, Unit] =
```
Comment (JaroslavTulach) [inline lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/file/BlockingFileSystem.scala:93]: See! It is called `copy` here ;-)

**Review by JaroslavTulach — APPROVED** [review]
I'd choose different name for the message and I'd provide a way to specify name of the new project.

File: lib/scala/json-rpc-server-test/src/main/scala/org/enso/jsonrpc/test/JsonRpcServerTestKit.scala [hunk 4]
```diff
@@ -174,10 +174,12 @@ abstract class JsonRpcServerTestKit
     def fuzzyExpectJson(
       json: Json,
       timeout: FiniteDuration = 5.seconds.dilated
-    )(implicit pos: Position): Assertion = {
+    )(implicit pos: Position): Json = {
```
Comment (hubertp) [inline lib/scala/json-rpc-server-test/src/main/scala/org/enso/jsonrpc/test/JsonRpcServerTestKit.scala:177]: Shouldn't we leave `Either` here?

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/service/ProjectService.scala [hunk 5]
```diff
@@ -490,20 +517,17 @@ class ProjectService[
...
-
+    CovariantFlatMap[F]
+      .ifM(projectRepository.exists(projectName))(
+        ifTrue  = findAvailableName(projectName, 1),
+        ifFalse = CovariantFlatMap[F].pure(projectName)
+      )
+      .mapError(toServiceFailure)
   }
 
+  private def getNameForDuplicatedProject(projectName: String): String =
```
Comment (hubertp) [inline lib/scala/project-manager/src/main/scala/org/enso/projectmanager/service/ProjectService.scala:528]: What if there is already one that exists?

**Review by hubertp — COMMENTED** [review]
I'd like to see a test case when we try to duplicate the same project twice. Also wondering how does cloud PM behave in such scenario (local PM should be consistent).

File: docs/language-server/protocol-project-manager.md (hunk 2 above)
Comment (4e6) [reply docs/language-server/protocol-project-manager.md:770]: In the cloud, the project with the ` (copy)` suffix is created, and then the user can rename it.

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/service/ProjectService.scala (hunk 5 above)
Comment (4e6) [reply lib/scala/project-manager/src/main/scala/org/enso/projectmanager/service/ProjectService.scala:528]: This function returns the suggested project name, and then we use `getNameForNewProject` function that checks if the current name exists and returns the final name for the project.

File: lib/scala/json-rpc-server-test/src/main/scala/org/enso/jsonrpc/test/JsonRpcServerTestKit.scala (hunk 4 above)
Comment (4e6) [reply lib/scala/json-rpc-server-test/src/main/scala/org/enso/jsonrpc/test/JsonRpcServerTestKit.scala:177]: If the JSON does not match the expected `json` value, the assertion will fail and the function will fail with the assertion exception

**Comment by 4e6** [comment]
> Also wondering how does cloud PM behave in such scenario (local PM should be consistent).

The Cloud just keeps creating projects with the ` (copy)` suffix. In the local project manager we don't allow creating projects with the same name

**Comment by somebody1234** [comment]
> The Cloud just keeps creating projects with the ` (copy)` suffix.

i guess it would be good to discuss with @PabloBuchu about what should happen if we duplicate files multiple times? i know in some other places we do use ` (2)`, ` (3)` etc

(i guess that happens when uploading files with the same name though)

it's worth noting that it's probably a bad thing if cloud allows multiple projects with the same name, since cloud assets are accessed by path, and cloud paths depend on the asset name

**Pull request merged**